crates/proposer-client/src/lib.rs

```diff
@@ -0,0 +1,81 @@
+use std::fmt::Display;
+use std::sync::Arc;
+use std::time::Duration;
+
+use serde::{Deserialize, Serialize};
+use sp1_sdk::network::B256;
+use sp1_sdk::{NetworkProver, ProverClient, SP1ProofWithPublicValues};
+
+pub use crate::config::ProposerClientConfig;
+pub use crate::error::Error;
+use crate::rpc::ProposerRpcClient;
+
+pub mod config;
+pub mod error;
+mod rpc;
+
+#[derive(Clone)]
+pub struct ProposerClient {
+    rpc: Arc<ProposerRpcClient>,
+    prover_client: Arc<NetworkProver>,
+    proving_timeout: Duration,
+}
+
+impl ProposerClient {
+    pub fn new(config: ProposerClientConfig) -> Result<Self, error::Error> {
+        let proposer_url = format!("http://{}:{}", config.proposer_host, config.proposer_port);
+        let cluster_url = format!(
+            "http://{}:{}",
+            config.sp1_cluster_host, config.sp1_cluster_port
+        );
+
+        let network_prover = ProverClient::builder()
+            .network()
+            .rpc_url(&cluster_url)
+            .build();
+
+        Ok(ProposerClient {
+            rpc: Arc::new(ProposerRpcClient::new(&proposer_url)?),
+            prover_client: Arc::new(network_prover),
+            proving_timeout: config.proving_timeout,
+        })
+    }
+
+    pub async fn request_agg_proof(&mut self, request: Request) -> Result<ProofId, Error> {
+        self.rpc.request_agg_proof(request.into()).await?.try_into()
+    }
+
+    pub async fn wait_for_proof(
```

Comment:
This could be part of a `trait` bounded to the generic, thus mocking the call could be easier for tests.
This can be done in another ticket/PR of course.

Reply:
Will add generic